### PR TITLE
Set user

### PR DESCRIPTION
**Pull request by M-OC**
- extracts types, constants into separate files
- extracts 'track', 'getPropertyId' and some minor methods into separate files
- check for 'propertyId' no longer occurs within 'track' method
- 'setUser' no longer accepts 'undefined'
- 'setUser' now accepts `{user: {email: 'foo'}}` and `{email: 'foo'}`

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -420,7 +420,7 @@ module.exports = {
         'import/exports-last': 'off',
         'import/group-exports': 'off',
         'import/no-cycle': 'error',
-        'import/no-default-export': 'error',
+        'import/no-default-export': 'off',
```
Comment (M-OC) [inline .eslintrc.js]: We were already ignoring this rule elsewhere.

File: src/declarations.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /* @flow */
 
-import type { MuseGlobalType } from './types';
+import type { MuseGlobalType } from './lib/types';
```
Comment (songz) [inline src/declarations.js]: would it make sense to create a `types` folder and put `MuseGlobalType` in `index.js` for now until we have more types? When I think about `lib` usually I think helper libraries and things like that. I think `types` should be in a folder of its own.

File: src/lib/track.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    const user = {
+        ...config.user,
+        id: getUserIdCookie()
+    };
+
+    // remove null, undefined values
+    user.id || delete user.id
+    user.email || delete user.email
+    user.name || delete user.name
```
Comment (M-OC) [inline src/lib/track.js]: The above is new.

File: src/lib/get-property-id.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { getClientID, getMerchantID } from '@paypal/sdk-client/src';
+
+export const getPropertyId = ({ paramsToPropertyIdUrl }) => {
```
Comment (M-OC) [inline src/lib/get-property-id.js]: This file is unchanged.

File: src/lib/compose-cart.js [hunk 5]
```diff
@@ -26,3 +33,39 @@ export const removeFromCart = (items, currentItems = []) => {
 export const addToCart = (items, currentItems = []) => {
     return [ ...currentItems, ...items ];
 };
+// $FlowFixMe
+export const composeCart = (type, data) => {
```
Comment (M-OC) [inline src/lib/compose-cart.js]: This method now uses 'constants' imported above.  Otherwise unchanged.

File: src/lib/cookie-utils.js [hunk 6]
```diff
@@ -21,3 +22,12 @@ export const setCookie = (cookieName : string, cookieValue : string, expirationM
     const expires = `expires=${ d.toUTCString() }`;
     document.cookie = `${ cookieName }=${ cookieValue }; Path=/; ${ expires }`;
 };
+
+export const getUserIdCookie = () : ?string => {
```
Comment (M-OC) [inline src/lib/cookie-utils.js]: These methods were removed from 'tracker-component'. They have not been altered. (note: ONE_MONTH_IN_MILLISECONDS should probably be in constants)

File: src/lib/constants.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+/* @flow */
+export default {
+    'sevenDays': 6.048e+8,
```
Comment (songz) [inline src/lib/constants.js]: Personally, (nitpick) I think its easier to understand if you wrote it like this: `sevenDays: 1000*60*60*24*7`

😆

File: src/lib/track.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    // to whatever you want it to be based on the trackingType string and data object.
+    // This can be useful for testing purposes, this feature won't be used by merchants.
+    if (config.paramsToBeaconUrl) {
+        img.src = config.paramsToBeaconUrl({ trackingType, data });
+    } else {
+        img.src = `https://www.paypal.com/targeting/track/${ trackingType }?data=${ encodeData(data) }`;
+    }
+
+    if (document.body) {
+        document.body.appendChild(img);
```
Comment (songz) [inline src/lib/track.js]: Just noticed that we are not removing the image element. Could you add a TODO here to remove the image after it has loaded?

File: src/lib/track.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    if (!getUserIdCookie()) {
+        setRandomUserIdCookie();
+    }
+    const user = {
+        ...config.user,
+        id: getUserIdCookie()
+    };
+
+    // remove null, undefined values
+    user.id || delete user.id
```
Comment (songz) [inline src/lib/track.js]: this is impossible right? Because right above you are setting userIdCookie

File: src/lib/track.js (hunk 3 above)
Comment (songz) [inline src/lib/track.js]: I thought setting `email` and `name` field to `null` or `''` will reset it on our back end? We want these values to get passed down the pipeline to get processed as is, so maybe we should not delete these fields?

File: src/tracker-component.js [hunk 10]
```diff
@@ -4,150 +4,31 @@ import 'whatwg-fetch'; // eslint-disable-line import/no-unassigned-import
...
+    IdentityData,
+    JetloreConfig,
+    Config
+} from './lib/types';
+
+const {
+    accessTokenUrl,
+    storage,
+    defaultTrackerConfig
+} = constants;
```
Comment (songz) [inline src/tracker-component.js]: why not just `import { accessokenUrl .... } from './lib/constants'`

File: src/tracker-component.js [hunk 11]
```diff
@@ -214,59 +95,26 @@ const getJetlorePayload = (type : string, options : Object) : Object => {
 
 let trackEventQueue = [];
 
-const track = <T>(config : Config, trackingType : TrackingType, trackingData : T) => {
+export const clearTrackQueue = (config : Config) => {
+    // $FlowFixMe
+    return trackEventQueue.filter(([ trackingType, trackingData ]) => { // eslint-disable-line array-callback-return
```
Comment (songz) [inline src/tracker-component.js]: nitpick: much cleaner to use a forEach and then `return []` imo

File: src/tracker-component.js [hunk 12]
```diff
@@ -214,59 +95,26 @@ const getJetlorePayload = (type : string, options : Object) : Object => {
...
+        track(config, trackingType, trackingData);
```
Comment (songz) [inline src/tracker-component.js]: question, does `track` also fire off JL events?

File: src/tracker-component.js (hunk 11 above)
Comment (joshbeam) [reply src/tracker-component.js]: Agreed. This was Zach's original implementation, and looks like it's just moved in the file. Maybe we can add a TODO instead to change this method?

Comment (songz) [reply src/tracker-component.js]: 👍

Comment (M-OC) [reply src/tracker-component.js]: added

File: src/tracker-component.js (hunk 12 above)
Comment (M-OC) [reply src/tracker-component.js]: It uses a different 'tracker' method. See "trackEventByType" & "getJetlore".

File: src/lib/track.js (hunk 8 above)
Comment (M-OC) [reply src/lib/track.js]: Done.

File: src/declarations.js (hunk 2 above)
Comment (M-OC) [reply src/declarations.js]: Agreed, but I'd prefer to do that + additional refactoring in another PR if possible.

File: src/tracker-component.js (hunk 10 above)
Comment (M-OC) [reply src/tracker-component.js]: Each key would have to be exported separately in that case.

File: src/tracker-component.js (hunk 12 above)
Comment (songz) [reply src/tracker-component.js]: https://github.com/paypal/paypal-muse-components/issues/53

**Pull request merged**